semgrep-core/src/parsing/ast/AST_dockerfile.mli

```diff
@@ -0,0 +1,194 @@
+(*
+   Dockerfile AST type definition.
+
+   Language reference:
+   https://docs.docker.com/engine/reference/builder/
+
+   Extends AST_bash.
+*)
+
+(*****************************************************************************)
+(* Token info *)
+(*****************************************************************************)
+
+type tok = Parse_info.t
+type loc = Loc.t
+type 'a wrap = 'a * tok
+type 'a bracket = tok * 'a * tok
+
+(*****************************************************************************)
+(* AST definition *)
+(*****************************************************************************)
+
+type shell_compatibility =
+  | Sh (* Bourne shell or Bash *)
+  | Cmd (* Windows cmd *)
+  | Powershell (* Windows powershell *)
+  | Other of
+      (* name of the shell extracted from the command name
+         in a SHELL directive *)
+      string
+
+type var_or_metavar =
+  | Var_ident of string wrap
+  | Var_semgrep_metavar of string wrap
+
+type string_or_metavar =
+  | Str_string of string wrap
+  | Str_semgrep_metavar of string wrap
+
+(* $foo or something like ${foo ...} *)
+type expansion =
+  | Expand_var of (* FOO in ${FOO} *) string wrap
+  | Expand_semgrep_metavar of (* $FOO in ${$FOO} *) string wrap
+
+(* Fragment of a string possibly containing variable expansions
+   or semgrep metavariables. This is similar to what we do in AST_bash.ml. *)
+type string_fragment =
+  | String_content of string wrap
+  | Expansion of (* $X in program mode, ${X}, ${X ... } *) (loc * expansion)
+  | Frag_semgrep_metavar of (* $X in pattern mode *) string wrap
+
+(* Used for quoted and unquoted strings for now.
+   Must be created with the 'create_str' function to ensure that
+   consecutive literal fragments are collapsed.
+   Converting str -> unchecked_str is done with the type coercion operator :>
+*)
+type unchecked_str = loc * string_fragment list
+type str = private unchecked_str
+
+val create_str : unchecked_str -> str
```

Comment:
The relevant changes in the module interface are the 3 lines of code above. The requirement of doing `(x :> unchecked_str)` before being able to pattern-match on it is impractical because it can't be done when pattern matching against `Arr_string x`.